src/Polly.Shared/Retry/RetryPolicyStateWithSleepAsync.cs

```diff
@@ -10,16 +10,21 @@ namespace Polly.Retry
 {
     internal partial class RetryPolicyStateWithSleep : IRetryPolicyState
     {
-        private readonly Func<Exception, TimeSpan, Context, Task> _onRetryAsync;
+        private readonly Func<Exception, TimeSpan, int, Context, Task> _onRetryAsync;
 
 
-        public RetryPolicyStateWithSleep(IEnumerable<TimeSpan> sleepDurations, Func<Exception, TimeSpan, Context, Task> onRetryAsync, Context context)
+        public RetryPolicyStateWithSleep(IEnumerable<TimeSpan> sleepDurations, Func<Exception, TimeSpan, int, Context, Task> onRetryAsync, Context context)
         {
             _onRetryAsync = onRetryAsync;
             _context = context;
             _sleepDurationsEnumerator = sleepDurations.GetEnumerator();
         }
 
+        public RetryPolicyStateWithSleep(IEnumerable<TimeSpan> sleepDurations, Func<Exception, TimeSpan, Context, Task> onRetryAsync, Context context) :
+            this(sleepDurations, (exception, span, i, c) => onRetryAsync(exception, span, context), Context.Empty)
```

Comment:
@SteveCote Same here, I think we should do:

```(exception, span, i, c) => onRetry(exception, span, c)```

to avoid the extra closure over the `context` variable